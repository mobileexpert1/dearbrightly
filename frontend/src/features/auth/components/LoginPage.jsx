import Link from 'react-router-dom/Link';
import styled from 'react-emotion';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import {
  isProcessing,
  getLoginErrorMessage,
  getOtpIsRequired
} from 'src/features/auth/selectors/authenticationSelectors';
import { facebookAuthenticationRequest, logInRequest } from 'src/features/auth/actions/authenticationActions';

import { history } from 'src/history';
import { loadFbApi } from 'src/common/helpers/fbInit';
import MessageBannerAlert from 'src/common/components/MessageBannerAlert';
import Input from 'src/common/components/Input/Input';
import FbButton from 'src/common/components/FbButton';
import { getEnvValue } from 'src/common/helpers/getEnvValue';
import { CustomSpinner } from "src/common/components/CustomSpinner";
import { StandardBlueButton, StandardOutlineBlueButton } from 'src/common/components/Buttons';
import OTPValidation from 'src/features/otp/OTPValidation';
import { GTMUtils } from 'src/common/helpers/gtmUtils';
import { fontFamily, fontWeight } from 'src/variables';



const DEBUG = getEnvValue('DEBUG');

const Container = styled('div')`
  @media screen and (max-width: 768px) {
    width: 100%;
  }
  @media (min-width: 768px) {
    width: 50%;
  }

  display: flex;
  flex-direction: column;
  padding-top: 10px;
  padding-right: 10px;
  padding-left: 10px;
  margin: 0 auto;
  margin-bottom: 50px;
  font-family: ${fontFamily.baseFont};
  .fb-btn-wrapper {
    padding: 10px 15px;
  }
`;

const ButtonContainer = styled('div')`
  @media (min-width: 768px) {
    padding-left: 15px;
    display: flex;
    align-items: center;
    text-align: center;
  }

  @media screen and (max-width: 768px) {
    justify-content: space-evenly;
    flex-wrap: flex;
    padding-left: 15px;
  }
`;

const LoginPageTitle = styled('div')`
  text-align: center;
  font-size: 30px;
  font-family: ${fontFamily.baseFont};
  font-weight: ${fontWeight.bold};
`;

const LoginPageForm = styled('div')`
  @media screen and (max-width: 768px) {
    flex-wrap: wrap;
  }
  @media (min-width: 768px) {
    display: flex;
  }
  align-items: center;
  position: relative;
`;

const LoginPageForgotPassword = styled('p')`
  cursor: pointer;
  margin-left: 15px;
  display: inline;

  :hover {
    color: #577e92;
  }
`;

const HelpBlock = styled('div')`
  position: absolute;
  bottom: 7px;
  color: red;
`;

const FBButtonContainer = styled('div')`
  display: flex;
  justify-content: center;
`;

class LoginPageContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: '',
      submitted: false,
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.goToForgotPasswordPage = this.goToForgotPasswordPage.bind(this);
    this.handleFBLogin = this.handleFBLogin.bind(this);
    this.FB = window.FB;
    this.emailRef = null;
  }

  componentDidMount() {
    loadFbApi();
    setTimeout(() => {
      if (this.emailRef) {
        this.emailRef.click();
      }
    }, 1000)
  }

  handleChange(event) {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  }

  handleSubmit(event) {
    event.preventDefault();

    this.setState({ submitted: true });
    const { email, password } = this.state;

    if (email && password) {
      this.props.logInRequest({ email, password });
    }

    GTMUtils.trackCall('login-login_click');
  }

  goToForgotPasswordPage() {
    GTMUtils.trackCall('login-forgot_password_click');
    history.push('/forgot-password');
  }

  toggleErrorMessage = () => {
    this.setState({ loginErrorMessage: '' });
  };

  handleFBLogin = () => {
    window.FB.login(
      response => {
        if (response.status === 'connected') {
          this.props.facebookAuthenticationRequest(response.authResponse, null, false);
        }
      },
      {
        scope: 'email',
        return_scopes: true,
      },
    );
  };

  track = () => {
    GTMUtils.trackCall('login-create_account_click');
  };

  inputRefCb = input => {
    if (input) {
      this.emailRef = input;
    }
  }

  render() {
    const { email, password, submitted } = this.state;
    const { isProcessing, loginErrorMessage } = this.props;

    return (
      <CustomSpinner spinning={isProcessing} blur={true}>
        <Container>
          {!this.props.otpIsRequired && (<LoginPageTitle>Log in</LoginPageTitle>)}
          {loginErrorMessage && loginErrorMessage.length > 0 && (
            <MessageBannerAlert text={loginErrorMessage} color="danger" />
          )}
          {this.props.otpIsRequired ?
            <div>
              <OTPValidation />
            </div>
            : (
              <div>
                <FBButtonContainer className="fb-btn-wrapper">
                  <FbButton onClick={this.handleFBLogin} text="Continue with Facebook" />
                </FBButtonContainer>
                <form name="form" onSubmit={this.handleSubmit}>
                  <LoginPageForm onClick={this.toggleErrorMessage}>
                    <div className={`form-group${submitted && !email ? ' has-error' : ''}`}>
                      <Input
                        autoFocus
                        inputRefCb={this.inputRefCb}
                        title="Email Address"
                        required={true}
                        handleChange={this.handleChange}
                        type="email"
                        name="email"
                        value={email}
                      />
                      {submitted && !email && <HelpBlock>Email is required</HelpBlock>}
                    </div>

                    <div className={`form-group${submitted && !password ? ' has-error' : ''}`}>
                      <Input
                        title="Password"
                        required={true}
                        handleChange={this.handleChange}
                        type="password"
                        name="password"
                        value={password}
                      />
                      {submitted && !password && <HelpBlock>Password is required</HelpBlock>}
                    </div>
                  </LoginPageForm>

                  <ButtonContainer>
                    <StandardBlueButton horizontalPadding={50} active={true} type="submit">Log in</StandardBlueButton>
                  </ButtonContainer>
                  <div style={{ marginTop: 10 }}>
                    <LoginPageForgotPassword onClick={this.goToForgotPasswordPage}>
                      Forgot your password?
                    </LoginPageForgotPassword>
                  </div>
                  <ButtonContainer style={{ marginTop: 30 }}>
                    <Link to="/register" className="login-page-button">
                      <StandardOutlineBlueButton
                        maxWidth="max-content"
                        active={true}
                        onClick={this.track}
                      >
                        Create account
                      </StandardOutlineBlueButton>
                    </Link>
                  </ButtonContainer>
                </form>
              </div>
            )}
        </Container>
      </CustomSpinner>
    );
  }
}

const mapStateToProps = state => ({
  isProcessing: isProcessing(state),
  userState: state.user,
  loginErrorMessage: getLoginErrorMessage(state),
  otpIsRequired: getOtpIsRequired(state),
});

export default connect(
  mapStateToProps,
  { logInRequest, facebookAuthenticationRequest },
)(LoginPageContainer);
