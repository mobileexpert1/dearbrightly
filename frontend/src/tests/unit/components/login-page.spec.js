import React from 'react';
import { mount } from 'enzyme';
import sinon from 'sinon';

import LoginPage from 'src/features/auth/components/LoginPage';

const loginAction = () => {};

describe('LoginPage', () => {
    it('should render correctly', () => {
        const component = mount(<LoginPage />);
        expect(component).toMatchSnapshot();
    });

    it('should handle change and submit', () => {
        const component = mount(<LoginPage loginAction={loginAction} />);
        const instance = component.instance();
        const handleChange = sinon.spy(instance, 'handleChange');
        const handleSubmit = sinon.spy(instance, 'handleSubmit');
        expect(component.state('email')).toEqual('');
        expect(component.state('password')).toEqual('');
        expect(component.state('submitted')).toEqual(false);
        handleChange({ target: { name: 'email', value: 'user@email.com' } });
        handleChange({ target: { name: 'password', value: 'password123' } });
        expect(component.state('email')).toEqual('user@email.com');
        expect(component.state('password')).toEqual('password123');
        handleSubmit({ preventDefault: () => {} });
        expect(component.state('submitted')).toEqual(true);
    });
});
