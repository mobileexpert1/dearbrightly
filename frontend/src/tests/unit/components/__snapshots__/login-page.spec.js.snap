// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should render correctly 1`] = `
<LoginPage>
  <div
    className="login-page-container"
  >
    <div
      className="login-page-section"
    >
      <div
        className="login-page-title"
      >
        Log In
      </div>
      <form
        name="form"
        onSubmit={[Function]}
      >
        <div
          className="login-page-form"
        >
          <div
            className="form-group"
          >
            <Input
              handleChange={[Function]}
              name="email"
              required={true}
              title="Email Address"
              type="text"
            >
              <div
                className="input-container"
              >
                <div
                  className="input-title"
                >
                  Email Address
                  <span
                    className="input-required"
                  >
                    *
                  </span>
                </div>
                <input
                  name="email"
                  onChange={[Function]}
                  type="text"
                />
              </div>
            </Input>
          </div>
          <div
            className="form-group"
          >
            <Input
              handleChange={[Function]}
              name="password"
              required={true}
              title="Password"
              type="password"
            >
              <div
                className="input-container"
              >
                <div
                  className="input-title"
                >
                  Password
                  <span
                    className="input-required"
                  >
                    *
                  </span>
                </div>
                <input
                  name="password"
                  onChange={[Function]}
                  type="password"
                />
              </div>
            </Input>
          </div>
        </div>
        <div
          className="login-page-form"
        >
          <Button
            onClick={[Function]}
            text="Login"
            type="black"
          >
            <div
              className="black-button"
              onClick={[Function]}
            >
              Login
            </div>
          </Button>
          <div
            className="login-page-forgot-password"
            onClick={[Function]}
          >
            Forgot Password?
          </div>
        </div>
      </form>
    </div>
    <div
      className="login-page-section"
    >
      <div
        className="login-page-title"
      >
        New Customer
      </div>
      <div
        className="login-page-details"
      >
        Create an account with us and you'll be able to:
      </div>
      <ul
        className="login-page-ul"
      >
        <li>
          Check out faster
        </li>
        <li>
          Save multiple shipping addresses
        </li>
        <li>
          Access your order history
        </li>
        <li>
          Track new orders
        </li>
      </ul>
      <Link
        className="login-page-button"
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/register"
      >
        <a
          className="login-page-button"
          onClick={[Function]}
          style={Object {}}
        >
          <Button
            text="Create Account"
            type="clear"
          >
            <div
              className="clear-button"
            >
              Create Account
            </div>
          </Button>
        </a>
      </Link>
    </div>
  </div>
</LoginPage>
`;
